File: python/sorts/merge_sort.py
# Python implementation for Merge Sort  
def merge(arr, l, m, r): 
	len1 = m - l + 1
	len2 = r- m 

	# temp arrays 
	left = [0] * (len1) 
	right = [0] * (len2) 

	# Copy data to temp arrays
	for i in range(0 , len1): 
		left[i] = arr[l + i] 

	for j in range(0 , len2): 
		right[j] = arr[m + 1 + j] 

	# Merge the temp arrays back to arr
	i = 0	 	# Index of left
	j = 0	 	# Index of right
	k = l 		# Index of arr

	while i < len1 and j < len2: 
		if left[i] <= right[j]: 
			arr[k] = left[i] 
			i += 1
		else: 
			arr[k] = right[j] 
			j += 1
		k += 1

	# Copy the remainder of left to arr
	while i < len1: 
		arr[k] = left[i] 
		i += 1
		k += 1

	# Copy the remainder of right to arr
	while j < len2: 
		arr[k] = right[j] 
		j += 1
		k += 1

def merge_sort(arr,l,r): 
	if l < r: 

		m = int((l+(r-1))/2)

		# Sort first and second halves 
		merge_sort(arr, l, m) 
		merge_sort(arr, m+1, r) 
		merge(arr, l, m, r) 

File: c/Sorts/MergeSort.c
void merge(int *array, int left, int middle, int right) {
	int i, j, k;

	int n1 = middle - left + 1;
	int n2 = right - middle;

	int lArray[n1], rArray[n2];

	for (i = 0; i < n1; i++) 
		lArray[i] = array[left + i];
	for (j = 0; j < n2; j++)
		rArray[j] = array[middle + 1 + j];

	i = 0;
	j = 0;
	k = left;

	while (i < n1 && j < n2) {
		if (lArray[i] <= rArray[j]) {
			array[k] = lArray[i];
			i++;
		} else {
			array[k] = rArray[j];
			j++;
		}
		k++;
	}

	while (i < n1) {
		array[k] = lArray[i];
		i++;
		k++;
	}
	while (j < n2) {
		array[k] = rArray[j];
		j++;
		k++;
	}
}

void mergeSort(int *array, int left, int right) {
	if (left < right) {
		int middle = left + (right - left) / 2;
		mergeSort(array, left, middle);
		mergeSort(array, middle + 1, right);

		merge(array, left, middle, right);
	}
}
